### PR TITLE
Add block restriction to phase field actions that define variables

### LINKED ISSUE #23632: Block restriction for phase field actions
## Reason
As reported in https://github.com/idaholab/moose/discussions/23593, the phase field actions could use block restriction for multiphysics coupling

## Design
Add a block parameter
check if it s passed before setting it to variables created by phase field actions
Not needed for kernels-creating actions as much, because the block restriction can be added to the variable instead

## Impact
Easier use of MOOSE for single-input phase field multiphysics

### PR DESCRIPTION
**Pull request by GiudGiud**
Responding to https://github.com/idaholab/moose/discussions/23593

Closes #23632

### CODE REVIEW

File: modules/phase_field/src/actions/NonconservedAction.C [hunk 1]
```diff
@@ -97,7 +101,6 @@ NonconservedAction::act()
     InputParameters params1 = _factory.getValidParams(kernel_type);
     params1.set<NonlinearVariableName>("variable") = _var_name;
     params1.applyParameters(parameters());
-
```
Comment (dschwen) [inline modules/phase_field/src/actions/NonconservedAction.C:100]: I'd prefer it if unnecessary white space changes we reduced to a minimum

File: modules/phase_field/src/actions/PolycrystalVariablesAction.C [hunk 2]
```diff
@@ -75,6 +77,9 @@ PolycrystalVariablesAction::act()
 
       var_params.applySpecificParameters(_pars, {"order", "family"});
       var_params.set<std::vector<Real>>("scaling") = {_pars.get<Real>("scaling")};
+      if (isParamValid("block"))
+        var_params.set<std::vector<SubdomainName>>("block") =
+            getParam<std::vector<SubdomainName>>("block");
```
Comment (dschwen) [inline modules/phase_field/src/actions/PolycrystalVariablesAction.C:82]: Would this not work?

```suggestion
      var_params.applySpecificParameters(_pars, {"order", "family", "block"});
      var_params.set<std::vector<Real>>("scaling") = {_pars.get<Real>("scaling")};
```

File: modules/phase_field/src/actions/PFCRFFVariablesAction.C [hunk 3]
```diff
@@ -61,6 +62,9 @@ PFCRFFVariablesAction::act()
 
   var_params.applySpecificParameters(_pars, {"family", "order"});
   var_params.set<std::vector<Real>>("scaling") = {getParam<Real>("scaling")};
+  if (isParamValid("block"))
+    var_params.set<std::vector<SubdomainName>>("block") =
+        getParam<std::vector<SubdomainName>>("block");
```
Comment (dschwen) [inline modules/phase_field/src/actions/PFCRFFVariablesAction.C:67]:
```suggestion
  var_params.applySpecificParameters(_pars, {"family", "order", "block"});
  var_params.set<std::vector<Real>>("scaling") = {getParam<Real>("scaling")};
```

? and so on

Comment (GiudGiud) [reply modules/phase_field/src/actions/PFCRFFVariablesAction.C:67]: never seen that. I ll try

File: modules/phase_field/src/actions/NonconservedAction.C (hunk 1 above)
Comment (GiudGiud) [reply modules/phase_field/src/actions/NonconservedAction.C:100]: removed them from the first commit

File: modules/phase_field/src/actions/BicrystalBoundingBoxICAction.C [hunk 4]
```diff
@@ -72,9 +72,6 @@ BicrystalBoundingBoxICAction::act()
       poly_params.set<Real>("inside") = 0.0;
       poly_params.set<Real>("outside") = 1.0;
     }
-    if (isParamValid("block"))
```
Comment (GiudGiud) [inline modules/phase_field/src/actions/BicrystalBoundingBoxICAction.C:75]: if my understanding is correct (and I just checked also) the applyParameters above does the job of adding block

File: modules/phase_field/src/actions/PolycrystalVoronoiVoidICAction.C [hunk 5]
```diff
@@ -57,9 +57,6 @@ PolycrystalVoronoiVoidICAction::act()
     poly_params.set<MooseEnum>("structure_type") = "grains";
     poly_params.set<UserObjectName>("polycrystal_ic_uo") =
         getParam<UserObjectName>("polycrystal_ic_uo");
-    if (isParamValid("block"))
-      poly_params.set<std::vector<SubdomainName>>("block") =
-          getParam<std::vector<SubdomainName>>("block");
```
Comment (GiudGiud) [inline modules/phase_field/src/actions/PolycrystalVoronoiVoidICAction.C:62]: same here

**Review by GiudGiud — COMMENTED** [review]
@dschwen your review should be addressed

File: modules/phase_field/src/actions/DisplacementGradientsAction.C [hunk 6]
```diff
@@ -66,7 +69,9 @@ DisplacementGradientsAction::act()
   {
     InputParameters params = _factory.getValidParams("StrainGradDispDerivatives");
     params.set<std::vector<VariableName>>("displacement_gradients") = _displacement_gradients;
-    params.set<std::vector<SubdomainName>>("block") = {"0"}; // TODO: add parameter for this
+    if (isParamValid("block"))
+      params.set<std::vector<SubdomainName>>("block") =
+          getParam<std::vector<SubdomainName>>("block");
```
Comment (dschwen) [inline modules/phase_field/src/actions/DisplacementGradientsAction.C:74]:
```suggestion
    params.applySpecificParameters(parameters(), {"block"});
```

File: modules/phase_field/src/actions/PolycrystalRandomICAction.C [hunk 7]
```diff
@@ -53,12 +53,9 @@ PolycrystalRandomICAction::act()
     // Set parameters for BoundingBoxIC
     InputParameters poly_params = _factory.getValidParams("PolycrystalRandomIC");
     poly_params.set<VariableName>("variable") = _var_name_base + Moose::stringify(op);
-    poly_params.set<unsigned int>("op_num") = _op_num;
     poly_params.set<unsigned int>("op_index") = op;
     poly_params.set<unsigned int>("random_type") = _random_type;
```
Comment (dschwen) [inline modules/phase_field/src/actions/PolycrystalRandomICAction.C:57]: quite annoying that the parameter type is unsigned int here, but that's a separate issue

File: modules/phase_field/src/actions/Tricrystal2CircleGrainsICAction.C [hunk 8]
```diff
@@ -67,9 +67,7 @@ Tricrystal2CircleGrainsICAction::act()
     poly_params.set<VariableName>("variable") = var_name;
     poly_params.set<unsigned int>("op_num") = _op_num;
     poly_params.set<unsigned int>("op_index") = op;
-    if (isParamValid("block"))
-      poly_params.set<std::vector<SubdomainName>>("block") =
-          getParam<std::vector<SubdomainName>>("block");
+    poly_params.applySpecificParameters(parameters(), {"block"});
```
Comment (dschwen) [inline modules/phase_field/src/actions/Tricrystal2CircleGrainsICAction.C:70]:
```suggestion
    poly_params.applySpecificParameters(parameters(), {"op_num", "block"});
```

File: modules/phase_field/src/actions/Tricrystal2CircleGrainsICAction.C [hunk 9]
```diff
@@ -67,9 +67,7 @@ Tricrystal2CircleGrainsICAction::act()
     poly_params.set<VariableName>("variable") = var_name;
     poly_params.set<unsigned int>("op_num") = _op_num;
```
Comment (dschwen) [inline modules/phase_field/src/actions/Tricrystal2CircleGrainsICAction.C:68]:
```suggestion
```

**Review by dschwen — APPROVED** [review]

**Pull request merged**